cookbooks/cdo-varnish/test/shared/shared.rb

```diff
@@ -0,0 +1,372 @@
+    mock_response url, text_cookie, {'Cookie' => 'cookie_key=cookie_value'}, {'Set-Cookie' => 'cookie_key2=cookie_value2; path=/'}
+
+    # Origin sees request cookie
+    response = local_request url, {}, {cookie_key: 'cookie_value'}
+    assert_equal text_cookie, last_line(response)
+
+    # Origin sets response cookie
+    response = local_request url
+    assert_equal text, last_line(response)
+    refute_nil /Cookie: [^\s]+/.match(response)
+
+    # Cache strips request cookie and strips response cookie
+    response = proxy_request url, {}, {cookie_key: 'cookie_value'}
+    assert_equal text, last_line(response)
+    assert_nil /Cookie: [^\s]+/.match(response)
+
+    # Cache hit on changed cookies
+    assert_hit proxy_request url, {}, {cookie_key: 'cookie_value3', key2: 'value2'}
+  end
+
+  it 'Allows whitelisted request cookie to affect cache behavior' do
+    url = build_url 5
+    cookie = "_learn_session_#{ENVIRONMENT}"
+    text = 'Hello World!'
+    text_cookie = 'Hello Cookie 123!'
+    text_cookie_2 = 'Hello Cookie 456!'
+    mock_response url, text
+    mock_response url, text_cookie, {'Cookie' => "#{cookie}=123;"}
+    mock_response url, text_cookie_2, {'Cookie' => "#{cookie}=456;"}
+
+    # Cache passes request cookie to origin
+    response = proxy_request url, {}, {"#{cookie}" => '123'}
+    assert_equal text_cookie, last_line(response)
+    assert_miss response
+
+    # Cache miss on changed cookies
+    response = proxy_request url, {}, {"#{cookie}" => '456'}
+    assert_equal text_cookie_2, last_line(response)
+    assert_miss response
+  end
+
+  it 'Strips non-whitelisted request cookies' do
+    url = build_url 6
+    cookie = 'random_cookie'
+    text = 'Hello World!'
+    text_cookie = 'Hello Cookie!'
+    mock_response url, text, {}
+    mock_response url, text_cookie, {'Cookie' => "#{cookie}=123;"}, {'Set-Cookie' => "#{cookie}=456; path=/"}
+
+    # Request without cookie
+    response = proxy_request url
+    assert_equal text, last_line(response)
+    assert_miss response
+
+    # Cache strips non-whitelisted request cookie and hits original cached response
+    response = proxy_request url, {}, {"#{cookie}" => '123'}
+    assert_equal text, last_line(response)
+    assert_hit response
+  end
+
+  it 'Strips non-whitelisted response cookies' do
+    # TODO: not implemented
+    skip 'Not implemented'
+
+    url = build_url 7
+    cookie = 'random_cookie'
+    text = 'Hello World!'
+    mock_response url, text, {}, {'Set-Cookie' => "#{cookie}=abc; path=/"}
+
+    # Response should have cookie stripped
+    response = proxy_request url
+    assert_equal text, last_line(response)
+    assert_nil /Set-Cookie: [^\s]+/.match(response)
+    assert_miss response
+
+    # Response should be cached even if response cookie is changed
+    mock_response url, text, {}, {'Set-Cookie' => "#{cookie}=def; path=/"}
+    assert_hit proxy_request url
+  end
+
+  it 'Does not strip cookies from uncached PUT or POST asset requests' do
```

Comment:
I believe you're understanding correctly, CloudFront will always perform the same cookie-filtering behavior for PUT/POST requests as it does for GET/HEAD requests for a specific path. Another way of putting it is that CloudFront does not consider the the request method before performing its cookie munging based on the provided whitelist.

Since there is no way to configure this differently as far as I've been able to tell, the only way to make CloudFront leave a cookie alone for PUT/POST request methods is to use a path with behavior that leaves a cookie alone for all request methods.

I don't think this is as important a behavior change as it might seem. We only modified our Varnish configuration to handle non-GET/HEAD requests in this way extremely recently (only 10 days ago in #4249, while PUT was handled as a special case 21 days ago in #4055). I've been in touch with @davidsbailey about the implications of this behavior change, and we've already committed a workaround (91e28c0e9ad8ddbd8bd022883f832cb26b63fc9f) that handles the one case that depended on this method-specific behavior.

I think it's a good idea to change the Varnish config to match what's capable in CloudFront, for this case (basically it would rollback #4055/#4249 in the updated VCL). That would be a quick update and shouldn't affect our application's current behavior at all.